### DISCUSSION TITLE
Add `text` platform

### DISCUSSION THREAD

**Discussion by raman325** [home-assistant/architecture#812] [Entity Models] [answered]
## Background

All input_* have corresponding platforms for integrations to use except `input_datetime` and `input_text`. By creating a `text` platform, integrations can start creating entities for different use cases (see #811 for `datetime` proposal). Examples include:
- Getting and setting display text on e.g. an lametric clock or a split flap display
- let a custom integration like keymaster provide text entities to set/view PINs on locks

## Proposal

Create a new `text` platform

### State

raw text

### Extra State Attributes

- `mode`: one of [text, password] defaulting to text, password affects how the UI displays the entity state as well as the UI for setting a new state
- `min`: min length of text defaulting to 0
- `max`: max length of text defaulting to 100 (capped at 255 by platform)
- `pattern`: pattern that string must match, both client side and server side defaulting to None

### Services

- `text.set_value` - Sets a new value for the text

### Integration must implement

- `native_value` is set by the integration. This allows the state to enforce min length, max length, and pattern
- `async_set_value` or `set_value` method which has a single string value as a parameter.

### Integration can implement

- `mode` property
- `native_min` property
- `native_max` property
- `pattern` property

### Implementation

https://github.com/home-assistant/core/pull/79454

**Comment by tsvi** [comment]
I might have completely misunderstood, but if we add `language` to the extra state attributes, that might allow the UI to pull the translation value accordingly from the integration. Maybe `set_value` should also support a language value as an argument?

**Reply by elupus** [reply]
I dont think translations make sense for this type of sensor.

**Comment by farmio** [comment]
I could use this well for the Knx integration. There are lots of wall mounted switches providing small (and some even big) LCD screens. Nearly all of them support display of custom text. 
This is already supported by leveraging a combination of string sensors and the notify platform - which is not really a user friendly solution. A text platform would be straightforward as, just like the number platform, these can be set and read from outside HA. 

The used datapoint txpes for that are DPT 16.000 ASCII and 16.001 ISO 8859-1 (Latin1) and Support up to 14 characters.

**Comment by elupus** [comment]
Esphome should be able to use this too for lcd screens and such.

**Comment by thecode** [comment]
I think this would be useful for many integrations to set device name for example, I am code owner for Switcher, the device has API to set the name, however it is not a REST/web API so it is only possible to do it with the manufacture app or via the API, having an option to set it from HA will be nice.

We do need to think how this is implemented, as I would prefer to avoid the need to combine it with another button platform, to set the value, so for example when a user click on the textbox to edit the name, would be better to also having a button to set the value so the user can edit and press "SET" (and we should also think if we the integration should support different button types such as SET/READ/WRITE to improve user experience)

**Reply by raman325** [reply]
I followed along with this until you mention set/write/read. I imagine the UI will be similar to input_text in that you type in a value and hit enter to set the new state. Does that address your concern, and if not, can you clarify?

**Reply by gjohansson-ST** [reply]
Don't see why this should work any different than the others e.g. number etc?

**Reply by thecode** [reply]
> I followed along with this until you mention set/write/read. I imagine the UI will be similar to input_text in that you type in a value and hit enter to set the new state. Does that address your concern, and if not, can you clarify?

Yes it does, enter is also ok

**Comment by KevinCathcart** [comment]
Conceptually I don't mind this. It is nicer than an integration needing to create helpers on the user's behalf, or similar. Admittedly keymaster was originally designed to be more "generate automations and helpers" that give a final result, and less a full integration, but I'd certainly slightly prefer if it could be using entities rather than helpers just because the helpers really clog up lists (the entities would still clog up entity lists, but that is better than clogging up both entity and helper lists).

I do think the naming is a bit unfortunate (just like with "number" and "boolean").

The issue being that entities of this platform are designed for input-only or bidirectional data, rather than for read-only data (which should use sensor platform entities, or binary_sensor for boolean), but the name is unable to indicate this, since the helper took the input version of the names. However, a "number" platform intuitively sounds like the right one for a numeric sensor value, and "text" for a textual one. 

I don't find "button" bad, since an always read-only button does not really make sense (albeit a sometimes disabled button is perfectly sensible).

I do think the ship has probably sailed here given past precedent, and this possible confusion is not really a big deal, but I thought it deserved to be mentioned. Obviously this same comment applies equally in the datetime context too.

**Reply by Shulyaka** [reply]
The original name for the `number` component was `analog_switch`, which is even worse:)

**Comment by gjohansson-ST** [comment]
I was looking at the selector so should we also consider multiline and suffix?

**Comment by frenck** [comment]
We have discussed this entity platform type in the past before, and from what I remember: Its ok to implement. At least, I'm ok with it 👍 

The thing missing in the proposal/discussion at this point: How would this look and work in the frontend? There are some concerns for the input: Single line? Multiline? Code highlighted? When does it confirm/apply/change state? On enter? On loosing focus of the field?

I think those are things that should be solved/discussed before opening PRs.

**Reply by raman325** [reply]
Yes I didn't touch the front end piece of this, but I think to start we should use the same control that's used for 'input_text' since it's already well understood and reduces implementation effort. I don't believe it supports multi line but we can reevaluate the need after initial implementation

**Reply by KevinCathcart** [reply]
Certainly if multi-line capability is something we determine is needed, it would also be beneficial for input_text too, if for no other reason than to enable users creating automations to set a text entity to one of two or more different values given via input_text helpers, based on events or computation or whatever. Such input helpers would need to match the capabilities of target text entity.

Thus I think it makes sense to defer until a need is shown, and then enhance both at the same time, right?

**Reply by gjohansson-ST** [reply]
It seems as a good idea to make it as ìnput_text`
The cases of multiline should be limited.
Could consider to add options for prefix and suffix given it's easy to implement and optional from the very start

**Comment by frenck** [comment, marked as answer]
✅ Looks good to go to me 👍

**Comment by frenck** [comment]
Hmm last, minute thought...

Should it be read-only too? As in, should set it to be an entity feature flag?
E.g., like the `update` entity, it can provide just a state, and can additionally provide the capability to set it...

Right now, we mainly use `sensor`s for that... but... should sensors become numeric only over time?

This thought might be quite a big shift... but, it might be the time to implement such a thing from the start?

**Reply by tsvi** [reply]
If `sensor`s should be numeric only, we need something to show textual values. Or automatically making a text into an enum?

**Reply by frenck** [reply]
I'm not understanding your comment @tsvi. My comment was about the `text` platform proposed here.

**Reply by tsvi** [reply]
I was relating to the statement "should sensors become numeric only over time?".

**Reply by frenck** [reply]
Ok, still not following. 

> If sensors should be numeric only, we need something to show textual values.

That is what this proposal is about? Introducing a `text` platform.

> Or automatically making a text into an enum?

Not sure what you goal is with that comment?

**Reply by tsvi** [reply]
Ok, now I get it. Didn't follow the full proposal. Sorry,

**Reply by KevinCathcart** [reply]
One thing that might be a bit unusual here architecturally is that for the most part we have kept entities that simply provide information that the user cannot attempted to set/control (sensor, binary_sensor, weather) separate from those where the user has services that attempt to influence the state: switch, light, lock, cover, number, etc.

The main entity types where not supporting any of the state changing services would quite normal seem to be camera (not all cameras support being turned on/off, or support native motion tracking), and update entities (which are encouraged but not required to implement the service to initiate updates).

This idea on the other hand would entail many text entities that are purely read-only, a few others that are read-write. Not necessarily a problem, but just a bit different from most other entity types.

The big question would be if such a migration provides enough benefit to be worth it. Migrating textual style sensors over to a new entity type (with a different domain) would be extremely disruptive to end users. I've seen plenty of users express significant annoyance about the amount of breaking changes requiring them to adjust their automation etc. And most of the changes people are complaining about are less disruptive than separating out non-numeric sensors to a new domain.

Now would such a separation bring benefits? Obviously yes it would. Would the benefits be big enough to justify a transition that would be as disruptive for end users as this would be? That is not as obvious, at least to me.

**Reply by frenck** [reply]
Extensively discussed this with some people, we are not going to pull the sensor entity apart right now.

So, the proposal is good to go 👍
